### PR TITLE
[JSC] Make TypedArray sort C++

### PR DESCRIPTION
**Pull request by Constellation**
#### a09e68d41ee57705c2f983578869cf4d72964372
<pre>
[JSC] Make TypedArray sort C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=264789">https://bugs.webkit.org/show_bug.cgi?id=264789</a>
<a href="https://rdar.apple.com/118368873">rdar://118368873</a>

Reviewed by Mark Lam.

Now, CachedCall gets much faster, so this patch moves TypedArray#sort implementation from JS to C++.
This is particularly good for TypedArray since they can just use Vector to allocate temporary buffer
for merge sort. When comparator is not specified, we use std::sort. But when it is specified, we use
our own merge sort function since std::sort will crash when values get changed during sorting.

We first copy TypedArray content, sorting, and then writing back to the original TypedArray. This is
because TypedArray can be detached or resized while executing comparator. So, for now, we just copy them
all so that it makes our algorithm simplified. It would be possible that we should reconsider using
our own quick sort for large TypedArray and checking detaching / resizing state every time executing
comparator.

* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/TypedArrayPrototype.js:
(at):
(linkTimeConstant.typedArrayMerge): Deleted.
(linkTimeConstant.typedArrayMergeSort): Deleted.
(sort): Deleted.
(toSorted): Deleted.
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayView.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewInlines.h:
(JSC::JSGenericTypedArrayView&lt;Adaptor&gt;::sort):
(JSC::JSGenericTypedArrayView&lt;Adaptor&gt;::sortFloat):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
(JSC::typedArrayMerge):
(JSC::typedArrayMergeSort):
(JSC::genericTypedArrayViewProtoFuncSortImpl):
(JSC::genericTypedArrayViewProtoFuncSort):
(JSC::genericTypedArrayViewProtoFuncToSorted):
(JSC::genericTypedArrayViewPrivateFuncClone): Deleted.
(JSC::genericTypedArrayViewPrivateFuncSort): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::typedArrayProtoSort const): Deleted.
* Source/JavaScriptCore/runtime/JSTypedArrayViewPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSTypedArrayViewPrototype::finishCreation):
* Source/JavaScriptCore/runtime/JSTypedArrayViewPrototype.h:

Canonical link: <a href="https://commits.webkit.org/270691@main">https://commits.webkit.org/270691@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/d71e470ddfa3285ac6a911b7c14a45e62e845167

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26129 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4739 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27402 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28225 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/23924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6508 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2151 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26384 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/3596 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/22487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28801 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/23448 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/22767 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/23820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/23834 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/27390 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/25344 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3243 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/1443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/32788 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/4649 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/7106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/3710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3363 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/3565 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | |

### CODE REVIEW

File: Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h [hunk 1]
```diff
@@ -710,41 +713,198 @@ ALWAYS_INLINE EncodedJSValue genericTypedArrayViewProtoFuncToReversed(VM& vm, JS
...
+
+    bool isResizableOrGrowableShared = false;
+    Structure* structure = globalObject->typedArrayStructure(ViewClass::TypedArrayStorageType, isResizableOrGrowableShared);
+    ViewClass* result = ViewClass::createUninitialized(globalObject, structure, length);
+    RETURN_IF_EXCEPTION(scope, { });
+
+    const typename ViewClass::ElementType* from = thisObject->typedVector();
+    typename ViewClass::ElementType* to = result->typedVector();
+
+    memmove(to, from, length * ViewClass::elementSize);
```
Comment (MenloDorian) [inline Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h]: Why memmove?  Why not memcpy given that we're obviously creating a brand new TypedArray instance via `ViewClass::createUninitialized`?

File: Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h [hunk 2]
```diff
@@ -710,41 +713,198 @@ ALWAYS_INLINE EncodedJSValue genericTypedArrayViewProtoFuncToReversed(VM& vm, JS
...
+    if (LIKELY(callData.type == CallData::Type::JS)) {
+        CachedCall cachedCall(globalObject, jsCast<JSFunction*>(comparatorValue), 2);
+        RETURN_IF_EXCEPTION(scope, { });
+        result = typedArrayMergeSort(vm, src, dst, [&](auto left, auto right) -> bool {
+            auto scope = DECLARE_THROW_SCOPE(vm);
+
+            cachedCall.clearArguments();
+
+            JSValue leftValue = ViewClass::Adaptor::toJSValue(globalObject, left);
+            RETURN_IF_EXCEPTION(scope, { });
```
Comment (MenloDorian) [inline Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:799]: I'm surprised that this is legal for a lambda that's supposed to return a bool.

**Review by MenloDorian — APPROVED** [review]
r=me

File: Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h (hunk 1 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h]: Yeah, changed it to copyElements.

File: Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h (hunk 2 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:799]: `{ }` is uniform initialization, which can initialize various things, including bool etc. :)

**Pull request merged**